### ISSUE TITLE
Incorrect behavior NOT IN operator

### ISSUE THREAD

**Issue by EvgeniyAkhr** [ClickHouse/ClickHouse#46202]
How to reproduce:
table 'TestStore' with ENGINE = ReplacingMergeTree
Query:
```
select distinct app, day
from TestStore
where day = toDate('2022-12-06')
  and app in ('Test1', 'Test2')
  and (app,  day) not in (
        select app, day
        from TestStore lds
        where day = toDate('2022-12-06')
          and app = 'Test1'
      );
```

Expected result:
This query should return one row with app = 'Test2', but nothing return.
But if I add “= 1” to NOT IN condition like as 
```
select distinct app, day
from TestStore
where day = toDate('2022-12-06')
  and app in ('Test1', 'Test2')
  and (app,  day) not in (
        select app, day
        from TestStore lds
        where day = toDate('2022-12-06')
          and app = 'Test1'
      ) = 1;
```
then anything is ok.
 
Clickhouse version: clickhouse-server:21.7

**Comment by Algunenano** [comment]
2 things:
* Please try to share a reproducible test case. What's `TestStore` and what does it contain?
* 21.7 is unsupported. Please try using a supported release. Last one is 23.1

**Comment by EvgeniyAkhr** [comment]
Reproducible test case: 

1. create table:

```
create table TestStore
(
    application                  LowCardinality(String),
    day                          Date,
    version                      DateTime
)
    engine = ReplacingMergeTree(version)
        PARTITION BY day
        ORDER BY (app, day)
        TTL day + toIntervalDay(35)
        SETTINGS index_granularity = 8192, min_merge_bytes_to_use_direct_io = 10737418240;
```

2. Add some records. For example: 
```
INSERT INTO TestStore(app, day, version) 
SELECT 'Test1' as app, '2022-12-06' as day, now() as version
;
INSERT INTO TestStore(app, day, version) 
SELECT 'Test2' as app, '2022-12-06' as day, now() as version
;
```
3. Try to run query:
```
select distinct app, day
from TestStore
where day = toDate('2022-12-06')
  and app in ('Test1', 'Test2')
  and (app,  day) not in (
        select app, day
        from TestStore lds
        where day = toDate('2022-12-06')
          and app = 'Test1'
      );
```
Expected result:
This query should return one row with app = 'Test2', but nothing return.

**Comment by Algunenano** [comment]
Returns the expected result in master and in 22.8:

```
$ clickhouse local
ClickHouse local version 23.2.1.1.

Mordor :) create table TestStore
          (
              app                  LowCardinality(String),
              day                          Date,
              version                      DateTime
          )
              engine = ReplacingMergeTree(version)
                  PARTITION BY day
                  ORDER BY (app, day)
                  TTL day + toIntervalDay(35)
                  SETTINGS index_granularity = 8192, min_merge_bytes_to_use_direct_io = 10737418240;

CREATE TABLE TestStore
(
    `app` LowCardinality(String),
    `day` Date,
    `version` DateTime
)
ENGINE = ReplacingMergeTree(version)
PARTITION BY day
ORDER BY (app, day)
TTL day + toIntervalDay(35)
SETTINGS index_granularity = 8192, min_merge_bytes_to_use_direct_io = 10737418240

Query id: a968565b-6111-4328-b201-34f659e6b9d0

Ok.

0 rows in set. Elapsed: 0.009 sec. 

Mordor :) INSERT INTO TestStore(app, day, version) 
          SELECT 'Test1' as app, '2022-12-06' as day, now() as version
          ;

INSERT INTO TestStore (app, day, version) SELECT
    'Test1' AS app,
    '2022-12-06' AS day,
    now() AS version

Query id: 34d95b5d-fe44-46b6-ae5d-d82d6c953d11

Ok.

0 rows in set. Elapsed: 0.010 sec. 

Mordor :) INSERT INTO TestStore(app, day, version) 
          SELECT 'Test2' as app, '2022-12-06' as day, now() as version
          ;

INSERT INTO TestStore (app, day, version) SELECT
    'Test2' AS app,
    '2022-12-06' AS day,
    now() AS version

Query id: 8827ca9c-b3c5-4045-b604-46d1fa832699

Ok.

0 rows in set. Elapsed: 0.001 sec. 

Mordor :) select distinct app, day
          from TestStore
          where day = toDate('2022-12-06')
            and app in ('Test1', 'Test2')
            and (app,  day) not in (
                  select app, day
                  from TestStore lds
                  where day = toDate('2022-12-06')
                    and app = 'Test1'
                );

SELECT DISTINCT
    app,
    day
FROM TestStore
WHERE (day = toDate('2022-12-06')) AND (app IN ('Test1', 'Test2')) AND ((app, day) NOT IN (
    SELECT
        app,
        day
    FROM TestStore AS lds
    WHERE (day = toDate('2022-12-06')) AND (app = 'Test1')
))

Query id: b3cb3b1b-a1a7-4389-aa56-2b1eda11a9b1

┌─app───┬────────day─┐
│ Test2 │ 2022-12-06 │
└───────┴────────────┘

1 row in set. Elapsed: 0.010 sec. 

Mordor :) 
```

Please upgrade

**Comment by EvgeniyAkhr** [comment]
Thanks for your response. I'll try to upgrade

**Comment by den-crane** [comment]
@EvgeniyAkhr but the issue does not reproduce even with 21.7 https://fiddle.clickhouse.com/218ad993-1933-4829-8640-43a4bf1edf8c

It's more complicated.

Provide 

```
select version();

set send_logs_level = 'debug';

select distinct app, day
from TestStore
where day = toDate('2022-12-06')
  and app in ('Test1', 'Test2')
  and (app,  day) not in (
        select app, day
        from TestStore lds
        where day = toDate('2022-12-06')
          and app = 'Test1'
      );


select distinct app, day
from TestStore
where day = toDate('2022-12-06')
  and app in ('Test1', 'Test2')
  and (app,  day) not in (
        select app, day
        from TestStore lds
        where day = toDate('2022-12-06')
          and app = 'Test1'
      ) = 1;

```

**Comment by den-crane** [comment]
I see. It's because of TTL

**Comment by den-crane** [comment]
https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelogs/v21.8.11.4-lts.md

Backported in https://github.com/ClickHouse/ClickHouse/issues/28756: Fix NOT-IN index optimization when not all key columns are used. This fixes https://github.com/ClickHouse/ClickHouse/issues/28120. https://github.com/ClickHouse/ClickHouse/pull/28315 ([Amos Bird](https://github.com/amosbird)).

```
grep -R 28315 *
v21.10.1.8013-prestable.md:* Fix NOT-IN index optimization when not all key columns are used. This fixes [#28120](https://github.com/ClickHouse/ClickHouse/issues/28120). [#28315](https://github.com/ClickHouse/ClickHouse/pull/28315) ([Amos Bird](https://github.com/amosbird)).
v21.8.11.4-lts.md:* Backported in [#28756](https://github.com/ClickHouse/ClickHouse/issues/28756): Fix NOT-IN index optimization when not all key columns are used. This fixes [#28120](https://github.com/ClickHouse/ClickHouse/issues/28120). [#28315](https://github.com/ClickHouse/ClickHouse/pull/28315) ([Amos Bird](https://github.com/amosbird)).
v21.9.2.17-stable.md:* Backported in [#28688](https://github.com/ClickHouse/ClickHouse/issues/28688): Fix NOT-IN index optimization when not all key columns are used. This fixes [#28120](https://github.com/ClickHouse/ClickHouse/issues/28120). [#28315](https://github.com/ClickHouse/ClickHouse/pull/28315) ([Amos Bird](https://github.com/amosbird)).
```
